ext/reflection/php_reflection.c

```diff
@@ -6497,11 +6527,37 @@ ZEND_METHOD(ReflectionAttribute, newInstance)
 		RETURN_THROWS();
 	}
 
-	if (!zend_get_attribute_str(ce->attributes, ZEND_STRL("phpattribute"))) {
-		zend_throw_error(NULL, "Attempting to use class '%s' as attribute that does not have <<PhpAttribute>>.", ZSTR_VAL(attr->data->name));
+	if (NULL == (marker = zend_get_attribute_str(ce->attributes, ZEND_STRL("attribute")))) {
+		zend_throw_error(NULL, "Attempting to use class '%s' as attribute that does not have <<Attribute>>.", ZSTR_VAL(attr->data->name));
 		RETURN_THROWS();
 	}
 
+	if (ce->type == ZEND_USER_CLASS) {
+		uint32_t flags = ZEND_ATTRIBUTE_TARGET_ALL;
+
+		if (marker->argc > 0) {
+			zval tmp;
+
+			if (FAILURE == zend_get_attribute_value(&tmp, marker, 0, ce)) {
+				RETURN_THROWS();
+			}
+
+			flags = (uint32_t) Z_LVAL(tmp);
+		}
+
+		if (!(attr->target & flags)) {
+			zend_throw_error(NULL, "Attribute '%s' is not allowed here", ZSTR_VAL(attr->data->name));
```

Comment:
Can we make "here" more specific?

Reply:
Also `'%s'` should be be `"%s"` here and below.

Reply:
Error messages now state the target and the allowed targets of the attribute. The messages have been unified with the ones used during compilation.